_posts/2022-01-10-reassign-partitions.md

```diff
@@ -0,0 +1,539 @@
+Here is an example topic configuration. You can change it to suit your requirements.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaTopic
+metadata:
+  name: my-topic-two
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  partitions: 10
+  replicas: 3
+  config:
+    retention.ms: 7200000
+    segment.bytes: 1073741824
+    # ...
+```
+
+Finally, let us create a separate interactive pod for running the commands in.
+
+> You may ask why we are using a separate pod to run the commands? Can't we just use one of the broker pods?
+> The issue is that running commands from within a broker is not good practice.
+> Running any of the Kafka `/bin` scripts from within the broker container will start another JVM (with all the same settings as the Kafka broker).
+> This can cause disruption, including causing the container to run out of memory.
+> So it is always better to avoid running the commands from a broker pod.
+
+You can use the following command to create a pod called `my-pod`:
+
+```sh
+kubectl run --restart=Never --image=quay.io/strimzi/kafka:0.29.0-kafka-3.2.0 my-pod -- /bin/sh -c "sleep 3600"
+```
+
+Wait till the pod gets into the `Ready` state.
+Once it is ready, we're all set to work through the examples.
+
+## Example 1: Decreasing replication factor
+
+When we just created the `my-topic-two` topic we specified `spec.replicas` as 4.
+In hindsight that's actually more replicas than we need, so let's reduce it to 3.
+To do this we can simply remove the last replica from the list of replicas of each partition of that topic.
+
+Since we're operating on a single topic it's going to be easiest to use the `--generate` mode of `kafka-reassign-partitions.sh` and then edit that (either by hand or using a tool like [jq](https://stedolan.github.io/jq/)).
+
+### Creating a proposal `reassignment.json` file
+
+As we discussed above, this file will have the topics that we need to reassign.
+
+Let's create our `topics.json` file in a file in the current directory of our local machine:
+```json
+{
+  "version": 1,
+  "topics": [
+    { "topic": "my-topic-two"}
+  ]
+}
+```
+
+After creating this file, copy it to the interactive pod container since we will be running all of our commands from there:
+
+```sh
+kubectl cp topics.json my-pod:/tmp/topics.json
+```
+
+We can now start a shell process in our interactive pod container and run the command inside it to generate our proposal `reassignment.json` data.
+Let's start the shell process:
+
+```sh
+kubectl exec -ti my-pod /bin/bash
+```
+
+Now it's time to generate the `reassignment.json` file. 
+We will use the `kafka-reassign-partitions.sh` command to generate our proposal `reassignment.json` data:
+
+```sh
+kubectl exec -ti my-pod /bin/bash -- bin/kafka-reassign-partitions.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 \
+--topics-to-move-json-file /tmp/topics.json \
+--broker-list 0,1,2,3 \
```

Comment:
does the deployed cluster have just 3 brokers so `0,1,2` ?

Reply:
@ShubhamRwt now you fixed the cluster having 5 replicas, should this list be `0,1,2,3,4`?